### PR TITLE
Implement certificate compression 

### PR DESCRIPTION
**Pull request by hwh33**
Certificate compression is a TLS 1.3 extension defined in [RFC 8879](https://datatracker.ietf.org/doc/html/rfc8879).

Resolves https://github.com/getlantern/lantern-internal/issues/5089.  See that ticket for motivation (since this repo is public).

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,14 @@
+module github.com/getlantern/utls
+
+go 1.16
+
+require (
+	github.com/andybalholm/brotli v1.0.4
+	github.com/klauspost/compress v1.13.6
```
Comment (hwh33) [inline go.mod:7]: I needed the above dependencies (for Brotli and Zstandard compression respectively), so I added a go.mod file.

File: .travis.yml [hunk 2]
```diff
@@ -8,12 +8,6 @@ dist: trusty
 
 install:
   - go get -t ./...
-  - go get golang.org/x/lint/golint
-  # Install gometalinter
-  - go get github.com/alecthomas/gometalinter
```
Comment (hwh33) [inline .travis.yml:13]: This linter is deprecated and was causing errors in CI.  I tried the recommended replacement (golangci-lint), but it produced way too many errors.

**Comment by hwh33** [comment]
Generally I would be wary of making changes to the std lib logic in this package.  However, I feel okay about this as (1) it's fairly straightforward and (2) there is little room for subtle bugs - we either decompress the cert correctly or we don't.

**Comment by hwh33** [comment]
Another note: it would be complex to add a full test for this logic.  I only added a client-side implementation, but a proper test would require a local server with support for certificate compression.  I did add the new message to existing tests for message marshaling / unmarshaling.

File: handshake_client_tls13.go [hunk 3]
```diff
@@ -493,7 +516,12 @@ func (hs *clientHandshakeStateTLS13) readServerCertificate() error {
 		c.sendAlert(alertDecodeError)
 		return errors.New("tls: received empty certificates message")
 	}
-	hs.transcript.Write(certMsg.marshal())
+	// [UTLS SECTION BEGINS]
+	// Previously, this was simply 'hs.transcript.Write(certMsg.marshal())' (without the if).
+	if !receivedCompressedCert {
+		hs.transcript.Write(certMsg.marshal())
+	}
```
Comment (forkner) [inline handshake_client_tls13.go:523]: okay, in the other case this has been already handled looks like on line 500.

File: .travis.yml (hunk 2 above)
Comment (forkner) [reply .travis.yml:13]: possibly look or file upstream as well? is this ults added cruft?

**Comment by forkner** [comment]
Some related discussion in the standard lib (from one of the main contributors to quic-go). not sure why it fizzled out -- https://github.com/golang/go/issues/42967

**Comment by forkner** [comment]
This seems like a general improvement @hwh33 -- may be worth prepping and proposing a patch upstream if it's not relying on other fork changes?

**Comment by forkner** [comment]
> Another note: it would be complex to add a full test for this logic. I only added a client-side implementation, but a proper test would require a local server with support for certificate compression. I did add the new message to existing tests for message marshaling / unmarshaling.
> 
> I also tested this against a number of live sites, including a number which support certificate compression and a number which do not. I manually verified that we were parsing and handling the compressed certificate as expected.

In lieu of that, possibly add a note on how/where to manually test, or add a (default disabled) test that requires hitting external resources we expect to use certificate compression?

**Comment by forkner** [comment]
all looks reasonable to me by eyeball.

File: .travis.yml (hunk 2 above)
Comment (hwh33) [reply .travis.yml:13]: Are you asking if this is the same as the upstream file?  If so, [yes](https://github.com/refraction-networking/utls/blob/master/.travis.yml)

File: handshake_client_tls13.go (hunk 3 above)
Comment (hwh33) [reply handshake_client_tls13.go:523]: Yes, that's correct

**Comment by hwh33** [comment]
> Some related discussion in the standard lib (from one of the main contributors to quic-go). not sure why it fizzled out -- golang/go#42967

Oh I'm surprised I didn't come across that while working on this.  I even read the [Brotli proposal](https://github.com/golang/go/issues/44444) which references that issue.  The 3x multiplier he references for QUIC handshakes is interesting.  Thanks for sharing.

> This seems like a general improvement @hwh33 -- may be worth prepping and proposing a patch upstream if it's not relying on other fork changes?

Good idea, I'll give it a shot!  Not sure how much Sergey keeps up with the upstream nowadays, but if anything it will be interesting to see if he responds =)

> In lieu of that, possibly add a note on how/where to manually test, or add a (default disabled) test that requires hitting external resources we expect to use certificate compression?

That's a good idea.  I'll go with a sort of mix and write a disabled test which can hit some flag-provided URL.

**Comment by hwh33** [comment]
Going to go ahead and merge this.  Thanks @forkner!

**Pull request merged**